packages/runtime/container-runtime/src/garbageCollection.ts

```diff
@@ -865,59 +877,92 @@ export class GarbageCollector implements IGarbageCollector {
+     * @param newRoutes - New references added by GC from the previous to the current run.
+     * @returns - a list of missing routes
      */
-    private validateReferenceCorrectness(currentGCData: IGarbageCollectionData) {
+    private findMissingRoutes(
+        currentGCData: IGarbageCollectionData,
+        previousGCData: IGarbageCollectionData,
+        newRoutes: Map<string, string[]>,
+    ): [string, string[]][] {
         assert(
-            this.gcDataFromLastRun !== undefined,
+            previousGCData !== undefined,
             0x2b7, /* "Can't validate correctness without GC data from last run" */
         );
 
-        // Get a list of all the outbound routes (or references) in the current GC data.
-        const currentReferences: string[] = [];
-        for (const [nodeId, outboundRoutes] of Object.entries(currentGCData.gcNodes)) {
-            /**
-             * Remove routes from a child node to its parent which is added implicitly by the runtime. For instance,
-             * each adds its data store as an outbound route to mark it as referenced if the DDS is referenced.
-             * We won't get any explicit notification for these references so they must be removed before validation.
-             */
-            const explicitRoutes = outboundRoutes.filter((route) => !nodeId.startsWith(route));
-            currentReferences.push(...explicitRoutes);
-        }
+        const getOutboundReferences = (nodeId: string, references: string[]): string[] => {
+            if (references === undefined) {
+                return [];
+            }
+            return references.filter((reference) => !nodeId.startsWith(reference) && isDataStoreNode(reference));
+        };
+
+        const getOutboundGraph = (routes: [string, string[]][]) => {
+            const outboundGraph: [string, string[]][] = [];
+            for (const [nodeId, references] of routes) {
+                const outboundReferences: string[] = getOutboundReferences(nodeId, references);
+                if (outboundReferences.length > 0) {
+                    outboundGraph.push([nodeId, outboundReferences]);
+                }
+            }
+            return outboundGraph;
+        };
 
-        // Get a list of outbound routes (or references) from the last run's GC data plus references added since the
-        // last run that were notified via `referenceAdded`.
-        const explicitReferences: string[] = [];
-        for (const [, outboundRoutes] of Object.entries(this.gcDataFromLastRun.gcNodes)) {
-            explicitReferences.push(...outboundRoutes);
+        // Get Current and Previous graphs
+        const currentGraph = getOutboundGraph(Object.entries(currentGCData.gcNodes));
+        const previousGraph = previousGCData.gcNodes;
+
+        // Expected Routes = Routes in the Current Graph and not in the Previous Graph
+        const expectedRoutes: [string, string[]][] = [];
```

Comment:
`expectedRoutes` does not seem like the right term for this. This is new references between previous and current run deteremined by generating the reference graph. You want to compare this with explicit references between previous and current run.